src/wolff_client_sup.erl

```diff
@@ -43,31 +47,57 @@ ensure_present(ClientId, Hosts, Config) ->
     {error, already_present} -> {error, client_not_running}
   end.
 
-%% ensure client stopped and deleted under supervisor
+%% @doc Ensure client stopped and deleted under supervisor.
 -spec ensure_absence(wolff:client_id()) -> ok.
 ensure_absence(ClientId) ->
   case supervisor:terminate_child(?SUPERVISOR, ClientId) of
     ok -> ok = supervisor:delete_child(?SUPERVISOR, ClientId);
     {error, not_found} -> ok
-  end.
+  end,
+  %% wolff_client process' terminate callback deregisters itself
+  %% but we make sure it's deregistered in case the client is killed
+  ok = deregister_client(ClientId).
 
-%% find client pid from client id
+%% @doc Find client pid from client id.
 -spec find_client(wolff:client_id()) -> {ok, pid()} | {error, any()}.
-find_client(ClientId) ->
-  Children = supervisor:which_children(?SUPERVISOR),
-  case lists:keyfind(ClientId, 1, Children) of
-    {ClientId, Client, _, _} when is_pid(Client) ->
-      {ok, Client};
-    {ClientId, Restarting, _, _} ->
-      {error, Restarting};
-    false ->
-      {error, no_such_client}
-  end.
+find_client(ClientID) ->
+    try
+        case ets:lookup(?WOLFF_CLIENTS_GLOBAL_TABLE, ClientID) of
+            [{ClientID, Pid}] ->
+                {ok, Pid};
+            [] ->
+                {error, no_such_client}
+        end
+    catch
+        error : badarg ->
+            {error, client_supervisor_not_initialized}
+    end.
+
 
+%% @private Make supervisor child spec.
 child_spec(ClientId, Hosts, Config) ->
   #{id => ClientId,
     start => {wolff_client, start_link, [ClientId, Hosts, Config]},
     restart => transient,
     type => worker,
     modules => [wolff_client]
    }.
+
+%% @doc Create a ets table which is used for client registration.
+%% Records are of format: {ClientId, Pid}
+create_clients_table() ->
+  EtsName = ?WOLFF_CLIENTS_GLOBAL_TABLE,
+  EtsOpts = [named_table, public, ordered_set, {read_concurrency, true}],
+  EtsName = ets:new(EtsName, EtsOpts),
+  ok.
+
+%% @doc Insert the client in the registration table.
+register_client(ClientId) ->
+    Pid = self(),
+    true = ets:insert(?WOLFF_CLIENTS_GLOBAL_TABLE, {ClientId, Pid}),
```

Comment:
Nit: maybe use `ets:insert_new` to avoid overwriting a previous entry?

Reply:
it should always overwrite.
client process may restart (by supervisor) after crash